### ISSUE TITLE
Plan for updating FsLab

### ISSUE THREAD

**Issue by dsyme** [fslaborg/zzarchive-FsLab#137]
Now that @zyzhu has started updating Deedle to netstandard 2.0 we should look at updating the whole FsLab collection

@zyzhu - do you use FsLab as an entire collection or just the individual pieces?

**Comment by dsyme** [comment]
Here is a proposed plan for updating FsLab

* [ ] Update packages to latest
* [ ] Add support for netstandard 2.0 (so you can reference FsLab in a netstandard 2.0 project)
* [ ] Update template and documentation to use `dotnet new`
* [ ] Update various things to assume .NET SDK is installed
* [ ] Check experience and tutorials on Mac, VSCode for Mac, VS for Mac

**Comment by zyzhu** [comment]
@dsyme I only use Deedle and XPlot. Data is cosumed via Dapper. I prototype all my research via FSI. I do hope dotnet core support on FSI will come soon as I refresh the issue everyday :)

I used to heavily use SqlClient type provider until it's not supported on dotnet core. I do use the latest SignalR so that Dapper can mitigate the problem.

I feel we shall discuss about a bigger picture of F# in data science, not only restrained by updating the current libraries to netstandard. @tpetricek 

Among FsLab packages, I found FSharp.Formatting as a weak link. I experimented using with FSharp.Formatting before but I found it's not too productive to use as it requires some boilerplate. Documentation and samples were not in good shape either. 

To data scientist or quant in finance, I generalize workflow as data retrieval -> computation -> visualization. Jupyter notebook is the de-facto place to start as almost zero boilerplate is required. The notebook can be easily rerun to replicate research step by step. On top of it, dynamic-typed language is very easy to get started for people with stats/math background. That's partially why Python build up an amazing ecosystem.

Though I strongly prefer strict-typed language, I cannot deny the productivity of Python ecosystem. But the productivity comes at a cost such as speed, scalability, type robustness and integration issues. But in practice, most research teams are willing to sacrifice these long-term cost with the boost of short-term productivity at idea generation stage.

In order to grow a community using F# in data science, avoiding boiler plate is a first prerequisite to boost productivity. The productivity from FSharp.Data typeprovider cannot make up the cost of preparing boilerplate to plot results. I found IfSharp is quite useful as I can run all my F# scripts and visualize results easily. The scripting and debugging experience is not as great as Visual Studio and FSI. But I just need to copy paste my script to IfSharp and use it as a notebook to record and visualize result. Most of time it works right away.

We shall improve the experience of FsLab on IfSharp as a priority by providing more documentations. It took me a while to dig through various issues to get a Deedle frame printed. Digging these issues will discourage newcomers to F#. More sample Azure notebooks on various topics to educate community will be very useful.

Another promising path is via a polyglot Jupyter notebook such as Beakerx by Two Sigma. It has just released version 1.0 with built-in two-way autotranslation. There was an old Beaker F# kernel. Maybe it could be ported so that F# can coexist with all other language/ecosystem on the same notebook. Then more libraries and visualizations can be handy.
https://github.com/twosigma/beakerx/issues/5039

Including a ML library such as Accord.Net/TensorFlowSharp is a good idea. But I am not an expert on it. Maybe @mathias-brandewinder has some good suggestions.

I would also include another optimization library Google.OrTools. They plan to release its FSharp library supporting netstandard in its next version soon. It can solve another branch of users on linear optimization cases. I've compiled its F# library to use in production and found its F# examples very elegant.
https://github.com/google/or-tools/issues/722

**Comment by TonyHenrique** [comment]
I don't know if I missed something, but I feel that F# need to support complex XSD -> XML generation. It is being used heavily here in Brazil by Governament for Sales, Medical, and it would be good to have a easy way to have type safety when generating XML from our data using the XSD Schemas provided by Governament.

See https://github.com/fsprojects/FSharp.Data.Xsd/issues/26#issue-336825128

**Comment by giuliohome** [comment]
My experience with data science (in the company where I work) is within a trading context, but (also in my personal programming experience) I can see the influence of machine learning (thinking about Kaggle) and big data tools (here I mean Hadoop hdfs and its vast ecosystem: so Kafka for example and Apache Spark especially for functional programmers).
It is worth noting that there is a strong usage of nosql in this world.
So I guess that machine learning is absolutely necessary in the mix and, furthermore, a basic sql interface is not sufficient (but a type provider, giving a greater sense of safety, on top of let's say a nosql db could be appreciated in many business use cases).
I love F# as a language and I'm sorry that Python is much more used in data science projects, but I would advice against interop with a dynamically typed scripting language; on the other hand the strategy of Scala can be instructive: building API for Apache Spark, but keeping its own distinctive functional programming approach as well. 
My 2 cents :)

**Comment by dsyme** [comment]
It's great discussion, please continue, all the comments are enlightening

Note the list of work items above is not meant to be comprehensive and is a bi stream-of-consciousness. 

My take is that FsLab should be a collection of packages which "work together and you don't regret".  That is, the packages should

* [ ] be useful for data science (but not necessarily a complete set of packages for every eventuality - you might need to add more)
* [ ] work cross-platform (including .NET Core)
* [ ] work with F# Interactive (including on .NET Core when it is done)
* [ ] work in iFSharp Jupyter notebooks
* [ ] be well-scoped, i.e. do what they say on the tin, and not more or less
* [ ] have relatively few bugs
* [ ] have an active maintainer
* [ ] be well-documented
* [ ]  be accepting contributions 
* [ ] not interfere with the use of alternative packages
* [ ] together they should not be "too large"
* [ ] be usable as independent components if necessary ("not a big fur-ball that is all or nothing")
* [ ] be usable in both data scripting and compiled code

Basically you want to to be able to "add an FsLab reference" and do some data-science workbook programming, whether that be in Visual Studio, VS Code, Jupyter notebooks or whatever.

Equally you should be able to back out of using FsLab and just use individual packages with the same effect.

Machine learning packages for .NET are a little tricky for FsLab.  The more complete ones like Accord.NET (which is great) tend to be a complete set of packages in their own right (which is also great). Other packages like ML.NET are a little too early to include.  So in general I agree 

Interop packages like RProvider, python , MATLAB provider, Excel provider etc. are tricky too. On the one hand these are incredibly useful when they are needed and work, and can benefit from regular integration and use with other components.  They are also sometimes painful to get working first time and people sometimes shy away from them.  On the other hand they are a source of considerable complexity and documenting them can be tricky.

Note that one approach would be to abandon FsLab as an "integrated" package and simply document the choices and how to get started with them

Finally FsLab today takes a very strong approach to literate programming - and I agree with @zyzhu that FSharp.Formatting is a bit of weak link.  I need to understand better where we should end up here.

**Comment by jackfoxy** [comment]
I think I'm close to having a netstandard2.0 PR
https://github.com/jackfoxy/XPlot/tree/magicmode
It builds in VS, but getting strange error with Newtonsoft.Json not recognized in the build target of the build script.

**Comment by dsyme** [comment]
@jackfoxy That's great.  I did a couple of updates to XPlot to fix the paket bootstrapper and documentation generation, you'll want to integrate those

**Comment by jackfoxy** [comment]
If it's already in master, I'll merge. I also implemented paket magic mode, which is possibly what you did @dsyme

**Comment by sebhofer** [comment]
Thanks to @zyzhu and @dsyme for starting this discussion. I'm really happy to hear about these developments! I agree with most of zyzhu's points; still, as one of your goals seems to be to attract new users to FsLab, I feel that providing my 2¢ of opinion could be helpful. From my experience starting out with using F# for data processing can be quite though for a newcomer (I'm coming from a science background) for several (some non-technical) reasons. Some thoughts:

- First and foremost: _getting to know what's available in FsLab_ is really though. Try clicking through the FsLab website and the project sites. Frankly, it's quite messy. Just looking at the projects listed on the top right is confusing; this top bar lists anything from 2 to 5 different packages, and hardly ever the same. To this day I don't know what exactly is "part" of FsLab. Also, some (but not all) pages link to http://fsharp.github.io/FSharp.Charting/, which is dead.
- Starting with Deedle was surprisingly hard for me coming from a dynamic world. Although the documentation is quite extensive, I still needed a lot of time to figure out seemingly simple tasks. I'm not quite sure how one could alleviate this. Maybe a list of common patterns in pandas and their translation to Deedle would help. I also thought about doing a Deedle cheat sheet along the lines of [the pandas one](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf), but I never got around to it.
- _Notebook interface_: I completely agree with @zyzhu that these are really useful, and I think it's crucial to have a notebook interface which nicely integrates display of dataframes and plotting without too much fiddling around.
- Interop with R and python would certainly be nice, and I think would attract many people who just can't afford to give up using a certain package for some reason or another.

**Comment by dsyme** [comment]
@sebhofer I agree with all those points, thanks My first aim here is to get FsLab "clean".  But we can also reassess its whole construction - I'm still not sure it should be anything but a template of the kind you propose (does it even need to be a combined nuget package?)

FsLab is, at the moment:
* Deedle
* XPlot
* Math.NET Numerics
* RProvider
* FSharp.Charting
* Some literate programming support

plus a template.  These seem reasonable and I think each is quite well documented (once links are all fixed). But the centrality of the literate programming support is questionable in the world of notebooks.

There are also transient dependencies on 
* Suave
* Newtonsoft.Json
* Google.DataTable.Net.Wrapper
the first two of which are questionable, and also optional dependencies on:
* Google charts
* Plotly
* R

**Comment by dsyme** [comment]
@jackfoxy Could you send a PR for your xplot .NET Standard 2.0 work, even if not yet quite complete?  Then we can discuss and others can help get it over the line? thanks

**Comment by jackfoxy** [comment]
@dsyme https://github.com/fslaborg/XPlot/pull/75

**Comment by cgravill** [comment]
I've merged a change to IfSharp to target .NET 4.7.1 to ease interaction with .NET Standard 2.0 https://github.com/fsprojects/IfSharp/issues/181 There is some odd behaviour but with that I'm able to use ML.NET 0.3 in the context of a Jupyter Notebook.

It'd be great to have improved support for FsLab. There was some initial work on this in https://github.com/fsprojects/IfSharp/issues/156 but more would be great. The helper script approach does have discovery issues but it's meant we can keep the core cleaner.

**Comment by sebhofer** [comment]
@dsyme I'm certainly in no position to judge the Deedle design, but I experienced that it's quite easy (for a beginner) to get bad performance if one is not careful. In my case I had to hack my own merge (or join?) function, because the built-in one would just not finish in reasonable time. (The reason was that the built-in version was too general for my problem and could be simplified considerably.) This is in principle not bad, but certainly slows you down in your day-to-day work. So there is certainly some room for improvement.

**Comment by aolney** [comment]
At the risk of piling on, since I was mentioned in an earlier post, I thought I'd give an update on Beaker for polyglot notebook programming. If this is an unfamiliar concept, basically it means you have a computational notebook that is simultaneously connected to multiple language kernels, and accordingly you can program in any of the corresponding languages across cells. So you can munge some data in F# and then in the next cell do some statistical modeling in R. 

I've been using Beaker for several years with real workloads, and it works very well. The project has recently pivoted towards supporting Jupyter, with a fairly huge loss of functionality during the pivot. The best current polyglot alternative within Jupyter seems to be the SoS kernel, [which can also be used in JupyterLab](https://github.com/vatlab/jupyterlab-sos). So far I've only used SoS for small workloads, but it seems very solid.

In Beaker I've had notebooks that use F#, R, Scala, Groovy, Java, and Javascript, using each where it works best (and potentially has a library dependency that I need). From my perspective, this is far better than trying to bring libraries developed in other languages into F# because:

- Native libraries are always current
- Native libraries have the best documentation/support
- Converted libraries can be more difficult/less fluent to use than native (sorry RProvider)

Polyglot notebooks can have some issues, but these seem to mostly be self-inflicted. For example, Beaker had specific kernel connection code for each supported language, making it difficult to maintain dozens of kernels. Also autotranslation (passing data structures between kernels through the notebook) is a cool and often touted feature that can be difficult to implement well with many edge cases. If autotranslation had a very basic implementation, then many of the associated problems would disappear. In practice I've found it's not really that useful except for passing configuration information between cells (e.g. file paths) because data of non-trivial size needs to hit the disk anyways, where it can be read by other kernels.

Anyways, it seems that polyglot notebooks are here to stay. I habitually use F# within this context (favorite language naturally) but use other languages in the notebook when their native support is a more natural fit. As far as F# kernels are concerned, since Jupyter has replaced Beaker, the ifSharp kernel is the best F# kernel to keep moving forward. I've used ifSharp with SoS/Jupyter and it works great.

**Comment by zyzhu** [comment]
@aolney Thanks for sharing your experience. You points clarify my confusion about Beaker and BeakerX. I took a quick look at SoS. It seems that it requires setting up a language module similar to https://github.com/vatlab/sos-r/tree/21883327750a1089066e8933843131d6271bfd74 so that SoS can interop F# with other languages. I found the documentation here https://vatlab.github.io/sos-docs/doc/documentation/Language_Module.html

Is that how you get started on using IfSharp on SoS? Any possibility to create a pull request to share your language module to SoS so that it can support IfSharp kernel out of box? That will help F# community get started on SoS notebook.

**Comment by aolney** [comment]
I've been using a [Jupyterlab installation](https://jupyterlab.readthedocs.io/en/stable/getting_started/installation.html) but I think the process is similar for Jupyter notebook.

I believe this is all it takes:

```
pip install jupyterlab
jupyter labextension install jupyterlab-sos
jupyter kernelspec install ifsharp 
```
In other words there's no need for a new language module b/c [ifSharp](https://github.com/fsprojects/IfSharp) works with Jupyter.

My understanding (could be wrong) is that the links you provided are only needed for certain functionalities like autotranslation and syntax coloring. They may also be needed for future capabilities like intellisense and linting.

**Comment by siavash-babaei** [comment]
So, say everything existing is updated and working nicely. Well, a proper RProvider would give you access to whatever is missing in F# and just about anything in python and then some. RProvider though has not worked for quite some time since R 3.5.

Some Notes and Long-Term Thoughts:

- **Data Frames**: For R and python (through pandas), data frames are kind of the primary core data structure that one would work with, especially in the case of in-memory data, and almost every library, package, and algorithm is aware of them and utilizes them in one way or the other. Going so far as even many big data tools simply chunk large data into manageable data frames and take it from there. On the other hand, no matter how good and effective Deedle is at handling data frames, Accord.NET, Math.NET Numerics, ML.NET, etc. are not aware of Deedle data frames and cannot directly consume them. (Please correct me if I am wrong ...) Hence severely limiting their usage.
- **Visualization**: In terms of visualizations, both R and python have superb capabilities in ggplot and Matplotlib. FSharp.Charting is not nearly as good. In addition, R has Shiny and python has Bokeh to handle interactive visualization and dashboard, etc similar to what Microsoft Power BI and Tableau offer in terms of preparing interactive reports and dashboards - although obviously less commercially polished ... So far as I know, F# has nothing similar. 
- **Data Retrieval**: FSharp.Data is great but it could be expanded upon perhaps incorporating some other TypeProviders as well maybe even supporting data file formats from R, Python, MATLAB, HDFS, SAS, SPSS. The ability to communicate with databases - Whether SQL or NoSQL - and various data sources, is of course of paramount importance. I have not seen TypeProviders for say MongoDB, Cosmos DB, HBase ... As an example, the last update on MongoDB.FSharp which is one of few such projects is 6+ years old and the link on MongoDB website is 8 years old: https://www.mongodb.com/blog/post/enhancing-the-f-developer-experience-with-mongodb.
- **Big Data**: I believe a significant proportion of data analytics workflows are still in-memory, using a standard panel of models for regression/classification/clustering tasks, and do not yet involve stuff like big data, Spark, Deep Learning, and alike. However, not having those capabilities, is a deal-breaker. Standard tools of the trade for that are Spark, Keras, Tensorflow, probably a couple of others would complete the list of necessary tooling. We got Spark for .NET and ML.NET is supposed to offer Tensorflow support, so I suppose the best hope lies with Microsoft and then maybe F# API/Wrappers in time. It's a pitty though ML.NET is written in C#, had it been F#, maybe it would have helped propel F# similar to what Spark did for Scala, not to mention that F# would have been a lot more suitable at core rather than OO C#. Nonetheless, ML.NET is an obvious candidate for inclusion in FsLab and bigger involvement from F# lads. It offers a very welcome more unified approach to doing data science plus support for Tensorflow and ONNX
- **Light Intuitive Syntax**: Since most of the time, you are doing exploratory analysis and prototyping, quick turn-around is a very important feature. What we have in the likes of ML.NET and Accord.NET is very C#, too awkward and verbose for quick and dirty hacking. In R, you would do:
```R
          model <- lm(data = scores, score ~ age * sex)
```
      and then, from this `model` object, you can extract whatever you need, including statistics, 
      coefficients and confidence intervals, error estimates, etc, even diagnostic plots, with some 
      pretty intuitive names. To me, doing the same thing as above and almost perfect in F# would 
      go like:
```FSharp
          let model = 
              let data = scores
              let response = [ "score" ]
              let predictors = [ "age"; "sex" ]
              (data, response, predictors)
              |> linearModel ModelType.OLS CrossEffects.Multiplicative
```
    with `model` object perhaps being a record type with fields corresponding to coefficients 
    table, error estimates, basic statistics, etc.

**Comment by siavash-babaei** [comment]
For whatever product, you would require a few killer features that would make it indispensable, and for F#, it could easily be the entire data analytics and data science workloads. The same thing that greatly helped propel python to the front. The user base, especially, being more mathematically inclined and comfortable with the syntax (I just love/adore it but dunno why but makes lots of people uncomfortable), ideas of immutability and the core of language being input -> function -> output, would be much better adopters than say, developers active in GUI or web. There are other areas I am sure, for example, business applications that fit nicely with Domain-Driven Design. But data science workloads - incidentally, a perfect match for DDD - are certainly worth the investment, especially as they seem to be exponentially growing both in volume and utilisation. If you think about it, one of the most active open source big data projects, Spark, is only 7 years old. The community seems to be more-so accepting of new tech that makes their life easier.
FsLab could be that unified environment for data analytics pipelines with a comprehensive suite of up-to-date tools accessible from whatever OS, with pieces and have the necessary awareness of each other. Kind of pointless to have a data frame that cannot be readily consumed within the tools that we use to analyse our data, say: ML.NET has its own data frame and btw, it seems very inferior to that of Deedle; and, Accord.NET has its own extremely horrible way of consuming data in the form of arrays.

**Comment by siavash-babaei** [comment]
Corporate support could be subtle, could be a lot of things from adoption and critique to code contribution to money, marketing, etc. For example, 

- a push to make F# code usable directly from within SQL Server and Power BI (remembering that a big component there, M Language, is inspired by F#) the same as is currently the case with R and Python; and/or, 
- a backend (or some form of help with one) for ML.NET and SPARK.NET in idiomatic F#.

**Comment by dsyme** [comment]
Adding comments from https://github.com/fslaborg/RProvider/issues/209

It's a good time to finally address this issue. There are many questions being discussed here. Let's just deal with the question of FsLab and its pieces.

Here are my opinions:

* The whole idea of a curated, unified collection like FsLab has turned out to be suspect as it doesn't really allow for change, evolution and deprecation unless very actively curated.

* The curation stopped because FsLab as a collection was based on .NET Framework, and some parts of the collection suffered badly in the transition to .NET Core. It only took one part to be still stuck on Mono or .NET Framework to render the whole thing stuck. That's what happened.

* With mono out of the loop things are easier once we establish a reasonable landing point

* MSFT is active in the parts of FsLab we directly care about - XPlot, fsdocs (which now generate .NET Interactive notebooks), .net interactive, F# literate scripting. It is also active in many related technologies. Other companies also contribute

* Reorienting to join forces with SciSharp, .net interactive and similar seems much more practical.

* FsLab certainly needs to be taken down and/or revamped on .NET Core only and/or wound up as a "one-stop shop technology". That will create space for better approaches I think. I'm open to suggestions but we need to rethink things.

* Note I'm not so interested in discussing this from a "future of F#" perspective (this has nothing to do with F# and web programming, for example) but rather just practical steps to get things cleaned up on on a good sustainable coherent basis going forward

Looking forward to making some progress here....

**Comment by dsyme** [comment]
BTW as can be seen from the discussion above I took a crack at modernizing FsLab to .NET Core in 2018.  I was honestly shocked how hard it was back then.  It will be much easier now.

FsLab had dependencies on FSharp.Formatting for literate scripting, including Razor for templating, and relied on mono.  At the time, FSharp.Formatting was barely functioning on .NET Core, and we only removed the Razor dependency earlier this year.  

Anyway, in 2020 I finally went through and added a usable .NET Command line tool to "fsdocs" to FSharp.Formatting which includes the literate scripting functionality of FsLab.

Another major factor is VS and VSCode.  VSCode is now the obvious place to centralise all such work.

**Comment by siavash-babaei** [comment]
ML.NET plus a feature-complete SciSharp with idiomatic F# support would solve most - if not all - problems: while C# syntax is awkward for data science pipelines, F# syntax sits very nicely with it. Dunno how doable that is, seeing as the current project appears to be a work-in-progress with some ways to go before all features are delivered. However, you would still definitely need a simple, easy, and quick way to interop with R or Python for cases not covered in SciScharp. 
@dsyme, FsLab becoming a "one-stop-shop technology", is actually the best approach. Who does not like convenience and more importantly, you gotta consider the users/customers of that tech which are a different breed of programmers/developers that the ones you would see in your usual software development industry, with different needs. Coding is not their primary purpose or activity, rapidly going through vast exploratory steps, discarding and selecting, and communicating results is.

**Comment by WalternativE** [comment]
I'm by no means a solid practitioner (currently trying to get into the field from my background as a software engineer) so I'm talking more opinion than knowledge here.

I think @siavash-babaei already lays down a lot of valid points. I'd just like to add, that it would be great to get an idea of where we are currently as a community in regards to data science/machine learning. Like the people at https://github.com/CSBiology who are using F# in their research and mainly fell beck to writing a lot of things themselves. SciSharp doesn't really have F# on their radar, at least it seems to me that way.

Many FSLab projects are indeed a bit 'stale'. I've been having good success in working with Deedle but there are a lot of thing one could improve (especially in writing docs). From a technology standpoint the new Microsoft data frame is really fancy (being built on Arrow) but I wouldn't have felt as productive as I did with Deedle. ML.NET works well enough with it in combination and even though the API is a bit odd it is quite fun to use - in the sense, that you feel quite productive. If you want to really tweak it, it gets kind of tedious.

The thing I'm missing most in comparison to a mature environment like the R Tidyverse is the ease of going from exploration to wrangling to modelling to validation and back again. In my current work with F# I really feel that the different libraries I use were built by entirely different teams with little to no regard for each other. Yeah you can somehow plot the data in your Deedle frame...you just have to 'un-deedle' it first. Same for modelling in ML.NET. It would be nice having a set of abstractions, that make this easier. At least some common ground with other projects. Like...I'm currently not even sure how many implementations of linear algebra libraries (and libraries that build upon them) are around. I just know, that they most likely aren't compatible. SAFE is great because it is a set of nice defaults. If I don't like the defaults it is trivial for me to swap them out.

Hope this added something other than more confusion to the discussion. 🙇‍♂️

**Comment by siavash-babaei** [comment]
Thank you @WalternativE. Actually, Tidyverse can act as a blueprint for similar capabilities in F#. The syntax is very nice and functional providing a sort of HowTo. The whole pipeline from importing, to cleaning, transformations, visualisation, modelling, and communicating results must be handled in ONE unified framework and just as importantly with a simplified, compatible, efficient, and intuitive syntax.

**Comment by zyzhu** [comment]
The visions discussed above sound wonderful. However, to be realistic, these visions needs a proper business plan and sponsorship for long-term sustainability.

RStudio is behind the huge push to modernize R development since Hadley Wickham joined RStudio in 2013. 
https://insights.stackoverflow.com/trends?tags=dplyr%2Cggplot2%2Ctidyverse
If you check a few tag trends on stackoverflow, Hadley's ggplot2 got popular since 2009, but tags such as `tidyverse` and `dplyr` only took off after 2016.

We already got clear indication from Don about what Microsoft team will focus on. I think the community shall just organically improve other individual components or create something from scratch. It is unrealistic to expect a holistic approach to build the whole data science ecosystem until a company like RStudio shows up. Maybe it never will show up.

Many magics in `tidyverse` relies on the flexibility of dynamic typing. I always feels bad seeing some `dplyr` samples like the following. To achieve similar result, it will be super verbose in `Deedle`
https://dplyr.tidyverse.org/index.html
```r
starwars %>% 
  mutate(name, bmi = mass / ((height / 100)  ^ 2)) %>%
  select(name:mass, bmi)
```
Where did `name`, `bmi`, `mass`, `height` come from? You don't need to declare anything but in R it just works as they are column names in `starwars`. 
Or this
https://genomicsclass.github.io/book/pages/dplyr_tutorial.html
```r
msleep %>% 
    group_by(order) %>%
    summarise(avg_sleep = mean(sleep_total), 
              min_sleep = min(sleep_total), 
              max_sleep = max(sleep_total),
              total = n())
```
This kind of exploratory work is very productive for a statisticians. Though the above lines are impossible to maintain, statisticians do not care as productivity of exploratory is the priority. I do not see any static typing language to beat this.

My point is that dynamic typing language and static typing language have their own pros/cons to exist in the world and we have to use both if we are in data science field.

**Comment by WalternativE** [comment]
Thanks for your insights @zyzhu. Yeah, I remember working in a small lab, that used R heavily before the RStudio people came along. It really wasn't as convenient as it is now (pretty chaotic if I remember correctly).

Regarding Microsofts involvement I see that interactive programming is getting a good push (both FSI and dotnet interactive which I've been using daily since it was stable enough to do so). FSharp formatting - or fsdocs if that's the better name now - is also coming along which is super important for documentation (one of the most important factors for me - I can go through source code but it will take me a while to get everything in my head). XPlot is already a library where I see a bit of friction. The CSBiology team already has a project which is currently living in the official plotly organization https://github.com/plotly/Plotly.NET. One can discuss about the API but it appears (at least from the Plotly side) more complete than XPlot. I'm also eagerly watchin the strides DiffSharp makes - one of the most exciting movements in the .NET space right now I'd say - and am all for backing the project. If there is a good ONNX story for DiffSharp I can totally see using interactive environments to work on models, special computing environments for training and ML.NET as a deployment target whenever a model goes into production (at least that is how it would work in my head - the usual disclaimer: I'm an enthusiast and no expert). Microsoft has its own ideas how to work with rectangular data in .NET. I'm not entirely sure how that's going to play out, I'm not even sure if Microsoft knows.

All of these projects are pretty disconnected from SciSharp. Basically every project I mentioned (apart from interactive programming and plotting) has one (or multiple) alternatives in this organization. From what I've experienced so far they aren't really compatible with most of it. If there were some abstractions we'd share I could see this change. Like in R it is really easy to work with most libraries because the notions of a vector, a matrix and a dataframe are engrained in everything. In the Scipy stack basically everything build on top of NumPy and whenever they use rectangular data they try to offer some sort of pandas interop. Right now - as I said before - we have quite a lot of linear algebra implementations, five - at least that's the amount I'd be aware of - major interpretations of rectangular data (three of them from Microsoft) and widely scattered statistics libraries. At least in the space of classical ML we're down to a narrow field because some of the old contenders simply died out. Compatibility between all of those components isn't really there so comitting to one takes away developer mindshare from the others. I love the F# community but we're simply not big enough to play zero sum games.

Relating to R being more flexible because it is a dynamic language. Yeah, that's right. The API is slicker due to not having to look out for types. As you mentioned, that makes working with it faster and the resulting analyses more brittle. I'm still pretty convinced, that it can be possible to get to a sweet spot where the API is - for the most part - statically typed and nice to work with. F# is capable of doing really impressive stuff with types but I'm totally with you, that it is not really possible to make it as convenient as the same API in a dynamic language. Still, I'm gladly taking the burden of writing more explicitly typed code if I can get some guarantees, that my code actually works (and keeps on working in the future). I don't want to bash R and/or Python but reproducing an environment using any of both languages is just so much pain. .NET is really good at working reliably (yeah - still my point even after the never ending story of moving to .NET Core, One .NET, whatever name they're going with today).

So, to finally get to a point. I can live with not having a one-stop-shop. I am happy with there only being a core of very well maintained libraries and tools, that help me to do DS/ML in .NET. I just think it would be necessary to have a group of people with skin-in-the-game, that talk about some foundational parts and interop. Some place to look to if you want to align with the rest of the "scientifically minded" .NET community. Everything, that enables individual contributors to build something without reinventing the whole stack from the ground up.

I hope my "ramblings" make some sense to you all and maybe contribute a bit to the greater discussion. 🙇‍♂️

**Comment by zyzhu** [comment]
Just to pile on a bit more. R can work with Python through `reticulate` package backed by RStudio. RStudio acknowledges the benefits of integrating two complementary ecosystem.
https://blog.rstudio.com/2019/12/17/r-vs-python-what-s-the-best-for-language-for-data-science/
https://rstudio.com/solutions/r-and-python/
I heard from @cartermp's video that the roadmap of dotnet/interactive might consider interop with Python. I just hope Microsoft has an even grander vision.

**Comment by siavash-babaei** [comment]
Dear @zyzhu, you are obviously right that there certainly are a great many serious hurdles, exacerbated by limitations of community size and corporate support. Then again, it is the question of the Egg and the Chicken: to expand community and attract support, you would need a working blueprint with bells and whistles and features to get some meaningful investment. The issues surely need carefully thought-out technical, business, and implementation plans but they are not by any means far-fetched or u attainable. Similar effort which went to SAFE Stack could just as easily and usefully turn into FsLab Stack. A lot of components are there like in SciSharp Stack or ML.NET but could definitely do with F# sugar which however you cut it, is much less involved than developing from scratch. 
While it is true that RStudio and Hadley Wickham with the addition of essentially Tidyvers dialect was a big boost for R, R was the biggest name in data science much before that, although the syntax was rubbish compared to now. In comparison, we have @dsyme and @tpetricek and then we got Microsoft which dwarfs RStudio by any measure. Microsoft has recently shown great willingness and initiative in data science from purchase of Revolution Analytics and turning it into Microsoft R, to addition of Python more-or-less as a first grade citizen and now bringing over of GVR. So they are willing to spend the money and contribute to development. We just need to carefully craft an approach and message,  utilize current assets within the company and take it from there.
I do not expect things to be as easy and concise in F# as is with dynamic languages but F# syntax is inherently more mathematical and less verbose than Python or R in many ways and trade offs in sound design and performance implementation and ... are well worth it. Frankly, I really liked Deedle and it offered a lot of tidyr and dplyr and lubridate functionality,  etc. I used Deedle to do a project ingesting, cleaning, and tidying large text files containing sensory data resulting in read-for-analysis CSV files. I did not experience problems, although few operations I needed to write myself but overall very nice and usable with easy and fluid syntax that actually reminded me a lot of R Tidyvers. For this project it was OK since the output was essentially a set of data frames or rather a data cube perhaps exported to files. It would have been problematic though had I needed to do analysis in F# since nothing else understands Deedle data frames. 
All in all, my experience is with F#, we can have a one-stop-shop, that even though strongly-typed, with its naturally concise and mathematical syntax, F# can still be very viable and pleasant when exploring data as well as much more  performant, safe, and easily maintainable when producing final results.

**Comment by nhirschey** [comment]
- **On Don's main point about FsLab as an organization**: agreed that a curated collection of packages doesn't work well when the underlying packages are not actively maintained. Deedle is the only one actively maintained. Even the main FSharp.Data is suffering and relying on Don for maintenance.
- **We need active maintainers**, especially ones in this thread coming from the engineering -> data perspective. This is the type of person who could help with things like getting the FSharp.Data build working again.
- **tidyverse**, what specifically is better? Tidyverse consists mostly of
    - Magrittr pipe operator `%>%` which was inspired by `|>` from F#. F# compares fine in this dimension.
    - dplyr filter/mutate/summarise verbs map very closely to the Seq/List/Array collections. Personally, I find base F# easier to understand than Deedle. Deedle is nice because of special functions for inequality joins, stats helpers that @zyzhu has added in, and you don't need to create as many record types as you iteratively reshape data (though anonymous records now help the type proliferation concern). But take @zyzhu's example, to my eye it is only slightly more verbose in base F#, though I admit being able to programmatically select record fields (as in `name:mass`) would be nice. 
    ```fsharp
     starwars
     |> Seq.map(fun x ->
       {| name = x.name; height = x.height; mass = x.mass; bmi = x.mass / ((x.height / 100.0)  ** 2.0) |})

     msleep
     |> Seq.groupBy(fun x -> x.order)
     |> Seq.map(fun (order, xs) ->
        let sleep_total = xs |> Seq.map(fun x -> x.sleep_total) |> Seq.toArray
        {| avg_sleep = Array.average sleep_total
           min_sleep = Array.min sleep_total   
           max_sleep  = Array.max sleep_total
           n = Array.length sleep_total |})
     ```
    - ggplot2 is great for publication quality graphics, but the plotly APIs seem fine with base F# collections (the equivalent of R vector, matrix, data.frame). Better control for saving Plotly images to files (like `ggsave`) would be useful. There's a 4 year old issue about this on XPlot (https://github.com/fslaborg/XPlot/issues/43).
- **Missing statistical modelling and model reporting**, this is the biggest issue for me. F#/C# has a good story for machine learning (due to Microsoft interest), but little in terms of statistics. This infrastructure is needed, but it's unclear to me whether to coordinate on extending ML.NET, Math dotnet numerics, the new Deedle regression interface, DiffSharp, etc. My sense is that there's little interest in this from the F# community.
  - There is no way to do panel data analysis (https://cran.r-project.org/web/packages/lfe/index.html or https://www.statsmodels.org/stable/index.html).
  - There is no way to do robust standard error estimation (https://cran.r-project.org/web/packages/sandwich/index.html or https://www.statsmodels.org/stable/index.html).
  - There is no way to generate publication quality model summaries (https://vincentarelbundock.github.io/modelsummary/).

**Comment by matthewcrews** [comment]
I would love to help with this project. I spent a lot of time in the R/tidyverse before becoming a full time F# dev. Having RStudio and a set of recommended libraries was a huge boost for early adoption. CRAN has a huge number of libraries that exist outside of the tidyverse that are still widely used (Ex: data.table). Having a prescription at the beginning streamlined people entering the ecosystem.

**Comment by matthewcrews** [comment]
I completely agree with @zyzhu when it comes to the productivity of Python/R. Dynamically typed languages have an "advantage" in that it is easy to start whipping together code quickly. Part of the reason I am so excited for the idea of Erased Discriminated Unions is that it address one of the pain points associated with a statically typed language compared to a dynamically typed ones. My hope is that we can build on this an potentially an enhancement to Computation Expressions to "re-capture" the productivity of R.

F# is a great language for DS/Analytics. I think with a few key enhancements, it could match the productivity of Python/R.

The Anaconda Python distribution solved the problem of "How do I get started?" I don't think F# should go that far, but having a curated list of recommended libraries and walkthroughs go a long way to streamlining adoption. The key metric I look at is, "How long does it take me to ingest an arbitrary CSV file, plot the data, and perform some kind of model fit?" Tightening that loop will be critical.

There are some "ergonomic" issues that could be improved with some key features that I think will make F# a more powerful language and better suited for Analytics/Modeling/Data Science.

**Comment by cartermp** [comment]
@WalternativE regarding this:

>  XPlot is already a library where I see a bit of friction. The CSBiology team already has a project which is currently living in the official plotly organization plotly/Plotly.NET. One can discuss about the API but it appears (at least from the Plotly side) more complete than XPlot

Since I'm the current maintainer of XPlot I'll say that I intend on people eventually moving over to Plotly.NET. We spoke with the team over at plotly and they're helping fund the effort. I think that XPlot is pretty good, but anything involving plotly should ultimately just use official bindings when a commercial entity like Plotly is interested in long-term maintenance. They were quite pleased with F# community activity in OSS and felt like it was a good investment.

So, where does that leave XPlot? Unsure, since the plotly package is the most-used and the most feature complete, and I'm likely to encourage people to move to Plotly.NET once it reaches 2.0.0. Its charter of having a consistent-ish API across different charting APIs remains unchanged and it's probably still a fine choice for several tasks.

Anyways, that's how I'd consider the issue of charting with Plotly as a backend moving forward.

**Comment by WalternativE** [comment]
@cartermp thanks for the update and thanks for your work in XPlot (and your work as our dearest PM while we're at it 🧡🧡🧡). It's wonderful to see, that there are multiple parties working in unison to make a stable data visualization library for .NET.

**Comment by cartermp** [comment]
@nhirschey and others - how is FSharp.Stats - https://github.com/CSBiology/FSharp.Stats - when it comes to being a good statistical package? I agree that this is one of the biggest gaps now, and I'm also not aware of any plans for Microsoft to publish anything in this space.

**Comment by siavash-babaei** [comment]
Dear @cartermp . See this example from 

```
// get coefficients of 3rd order regression polynomial 
let regressionCoefficients = Fitting.LinearRegression.OrdinaryLeastSquares.Polynomial.coefficient 3 x_Data y_Data

// get fitting function of 3rd order regression polynomial
let regressionFitFunc = Fitting.LinearRegression.OrdinaryLeastSquares.Polynomial.fit 3 regressionCoefficients 
```
A better design in my opinion, for example, would yield a single object from which one would then extract various bits per need. Furthermore, why would you unnecessarily expose users to technical underpinning - like the algorithm used for estimating model coefficients - they would not need most of time.

```
let regModel = Fitting.LinearRegression.OrdinaryLeastSquares.Polynomial.fit 3 x_Data y_Data

let regCoeffs = regModel.Coefficients
let regDiags   = regModel.DiagnosticPlots
let regEstms  = regModel.Estimates
...
```

**Comment by cartermp** [comment]
> Regression is a staple machine learning technique.

Oh for sure, and I think that the lines can sometimes get a little blurred between what is machine learning and what is plain old statistics. But I'm more curious if the library looks as if it could service some of the needs that @nhirschey is speaking towards. The shape of the API could change over time, especially if there is feedback indicating that it's conceptually challenging. I think the key thing is that there are APIs available, though, no matter how strange they may feel.

**Comment by lqdev** [comment]
Apologies ahead of time for the brain dump.

Love this discussion and would like to contribute in any way to help move this forward. Some of my (biased) thoughts: 

- Within .NET, F# is the language for analytical workflows. I, as I'm sure many of you do, strongly believe this to be the case.
- ML.NET has a strong value proposition when it comes to modeling and deployment. Many classical machine learning algorithms are implemented and performance especially on large datasets is something that's difficult to get in the Python/R ecosystem. Regarding deep learning, there is more limited support, but the extensibility of the framework through TensorFlow.NET and ONNX help overcome some of the limitations. From a deployment standpoint, it runs (almost) anywhere .NET runs. This to me is extremely important. For various reasons, model deployment is still a problem that's hard to solve. The typical approach is, let's wrap the model in a Flask web service and call it a day. I'm sure @matthewcrews may have a thought or two on the topic :). In .NET, even if you're deploying your model as a web service, from a performance standpoint, you get so much for free with ASP.NET core. Not to mention using something like Giraffe or Saturn on top of that and you get additional benefits. In .NET you can take deployments a step further and also deploy to Desktop, WASM, and hopefully Mobile / IoT devices (when ARM support is made available).
- With that said, ML .NET has its faults. Reconciling the first and second points, the ML .NET API is verbose and is weird to use from an F# standpoint. Not unusable, but also not fluid. There are a lot of things that happen before modeling from a data loading, wrangling and visualization standpoint. These are all areas where ML .NET falls short although many of the components are there from a .NET standpoint. There is the DataFrame API, .NET for Apache Spark and IDataView to represent data, but there's little to no interop between them. Same with visualizations. There is little to no interop that allows for seamless visualizations coming from these formats. I'm sure @WalternativE has plenty of thoughts on this :) . In the case of the missing parts of ML .NET and better F# support are areas that the ML .NET is actively looking into. I think this thread helps illustrate a lot of the pain points.
- Keeping in mind maintainability and resources, abstractions and extensibility may the path of least resistance as others have mentioned. Particularly areas where these efforts are built around standards. When it comes to data representation, there's [Apache Arrow](https://github.com/apache/arrow/blob/master/csharp/README.md). To a certain extent, Arrow is supported by .NET for Apache Spark DataFrames, DataFrame API, and I believe the IDataView as well. Greater interop between these in the .NET ecosystem as well as extensibility to libraries from other languages would make it easier to move back and forth between these. Likewise, from a visualization standpoint, using a standard data representation format may help interop not only within .NET but more broadly with other Python / R libraries. A bit off topic, but .NET Interactive has shown these polyglot type of environments may work well together. From a modeling standpoint and deployment standpoint, the same applies. Building around standards. ONNX is one of those standards. Although ONNX itself has some shortcomings, mainly around not fully delivering on the interop promise because in the end it's up to each framework to implement and maintain the ONNX operations. TensorFlow is one of the biggest offenders when it comes to not making it easy to use with ONNX. This is for inferencing purposes. However, for training, I believe there are some investments being made around training [Announcing accelerated training with ONNX Runtime—train models up to 45% faster](https://medium.com/microsoftazure/accelerating-model-training-with-the-onnx-runtime-519d75a97166).

I'm not sure that a one stop shop solution can address all of the individual steps in the workflow. For example, in Python, you do your data wrangling with Pandas or Spark, but maybe model with SparkML/Scikit/PyTorch. But having something like FsLab, or maybe now SciSharp as a collection or proposed SAFE stack for ML definitely is something worth doing. So long as these libraries embrace the concepts of interop and extensibility built around standards within the .NET ecosystem as well as the overall DS/ML ecosystem. I'm not sure if this is the best way to go about it. I've created a [Gist](https://gist.github.com/lqdev/3c3025f9a09971fb6a1121efa3e32cf5) with different steps in the data science / ML workflow steps and .NET libraries that provide support for it. I'm happy to take suggestions on where something like that may live as well as helping add to it to get the curation part going for maintained libraries.

**Comment by matthewcrews** [comment]
@lqdev I really appreciate hearing your thoughts and they echo many of the concerns that have come to my own mind. I think F# is well positioning to be a "full-workflow" language for Data Science/ML. This includes data wrangling, transformations, modeling, testing, and finally deployment. Python/R have optimized the Data Scientist workflow (from data import to validated model) but they are poor when it comes to putting a model into production. Python/R are great languages but they were not build with Production deployments in mind first and foremost (I'm sure I will catch grief for this). They are great languages, for sure, but the jump from development to production is non-trivial (runtime and library versions can be a real nightmare, not to mention debugging, logging, error reporting, refactoring, etc.).

I realize that my initial comments were off topic from what @dsyme  was originally asking for so my apologies. I believe Python/R have benefitted from a "standard" set of beginner libraries. 90% of an analysts work can be done with those initials libraries. If a beginners needs are not met with those initial libraries, it is easy to reach out to the community to find a more apt library (Ex: `data.table` for data munging). I think it would be worthwhile to have a curated list of "basic" libraries which covered the basic workflows for Data Science/Analytics in the F# stack.

Long term I think that we should consider someone advocating for the "ergonomics" for F# for Data Science/Analytics/Statistics. It goes beyond just having the necessary libraries. The workflow of an analyst is fundamentally different from a software developer. I don't think these things need to be at odds though. I think there can be a synthesis of needs. Much of what I do at Quicken these days is taking the needs of disparate departments and finding the solution that integrates these requirements. There are a couple of key features which could really close the gap with Python/R when it comes to analyst productivity.

I would love to help in easing the onboarding with new F# developers however I can. I'm working with my company to try to create an F# advocate position internally. We will see how that goes. We benefit a lot from the F# language so I hope to make the case we could make investing in the community a first class investment.

**Comment by nhirschey** [comment]
@cartermp, I wasn't familiar with FSharp.Stats before this thread. Looking through it, it has nice F# syntax for descriptive statistics like Seq.stdDev, etc. That is useful and they're building out a good foundation. I applaud their work. 

However from what I can tell, like ML.NET, MathNet.Numerics, Accord, etc., the regression modelling is currently inadequate for research purposes. There are no facilities for panel data models (https://en.wikipedia.org/wiki/Panel_data) or time series models with appropriate standard errors. I had hoped that ML.NET would develop these, but it still only has facilities for plain-vanilla standard error calculations in regressions. This is why despite preferring F# I always have to drop back into R whenever I get to the modelling stage of analysis.

I know this is a chicken/egg problem. And I'm thrilled with the F# 5.0 scripting and notebook improvements. I only point it out in the spirit of "what are some things holding back F# in the data science area."

**Comment by kMutagene** [comment]
Very interesting thread, lucky that i found out about it via twitter 😄. This post may get quite long, i hope that it can add some value to the discussion

Seems like the time where we from @CSBiology can chime in and try to contribute to this conversation. 

> Within .NET, F# is the language for analytical workflows. I, as I'm sure many of you do, strongly believe this to be the case.

This is our central dogma for 5+ years now. We are in our core a bioinformatics workgroup at a German university. We are using F# for data analysis and research. We love it's syntax, immutability by default, and that it has a functional-first approach. We believe that for these reasons it is perfect for the general workflows that we see everyday, where you basically have several (complex) mapping and folding steps of data to produce results. This is also the reason why we prefer it over both R and python, which are the de-facto 'standard' in our field due to majority vote.

When we started using F# extensively (and now almost exclusively), we saw a lack of, let's call them `standard libraries` for our kind of application, the most important ones being `plotting` and `(descriptive) statistics`.

Therefore, we were often forced to write our own functionality for specific tasks in our analysis scripts, which eventually resulted in libraries like [BioFSharp (our main bioinformatics library)](https://github.com/CSBiology/BioFSharp), [FSharp.Stats](https://github.com/CSBiology/FSharp.Stats), and FSharp.Plotly (which is now [Plotly.NET](https://github.com/plotly/Plotly.NET) directly maintained under the plotly organization). I am aware that alternatives for some of these libraries were developed in parallel, but the general workflow for data analysis that we are using is the following:

- Read raw data either via format-specific readers or Deedle (which is an awesome library by the way)
- Aggregating and transforming the data based on the question at hand, either by specific functions from BioFSharp or any kind of statistical functionality of FSharp.Stats. This happens on the Deedle frame when possible, but there may be the need to get the numeric columns from deedle, process the 2D array, and then go back into the Frame environment.
- visualize analysis performance indicators, data distribution, and results via Plotly.NET at any point of the analysis, use the interactive plot to get insights and repeat the analysis cycle accordingly.
- Write the result files and annotate the computational workflow via ontology annotations (admittedly WIP)

The reason why these libraries may be lacking functionality in specific regards or cover more than the name might suggest is that we improve them on a use-case basis, meaning as long as we don't need new functionality in our analysis pipelines, those features keep a low priority. FSharp.Stats is a good example here, as it contains e.g. pretty extensive fitting and signal processing features you might not expect from just the name, but may be lacking statistical features as indicated by other posts here.

What all of this basically boils down to is the following: we gladly contribute to the F# OSS community, but may not have the time and resources to make these libraries feature complete. We would however love to have more contributors and help shape the data science landscape in F# and .NET. This may also be by integrating our projects into existing ones.

**Comment by bvenn** [comment]
@cartermp @nhirschey
Our idea for FSharp.Stats as also laid out by @kMutagene emerged due to the reason you already mentioned. There was a lack in statistical packages compatible with our F# data analysis workflows. Step by step we aim to spot and implement basic, as well as higher level statistical functions/algorithms applicable in scientific data analysis. As @kMutagene already mentioned the project is extended when further functionalities are required in order to analyze our data.

FSharp.Stats aims to cover topics as: 
- fitting and interpolation
- signal processing and clustering
- linear algebra
- and of course descriptive statistics and statistical testing.

Regarding data fitting, FSharp.Stats covers [simple linear regression](https://csbiology.github.io/FSharp.Stats/GoodnessOfFit.html) with confidence and prediction band determination, [polynomial regression, and cubic/smoothing spline regression.](https://csbiology.github.io/FSharp.Stats/Fitting.html)
Models for nonlinear regression cover several [growth curve models](https://csbiology.github.io/FSharp.Stats/GrowthCurve.html) and common other models as gaussian, exponential or logistic functions.

If you have suggestions for panel data regression strategies or any other, we would be happy to consider it for future extension of the FSharp.Stats library.

**Comment by nhirschey** [comment]
@bvenn and @kMutagene, sounds good. I'll do some work with your library to get familiar and then raise some issues in your dev repo where we can discuss how to collaboratively improve the feature set and api ergonomics.

**Comment by kMutagene** [comment]
@cartermp

Regarding all things plotly (I maintain both Plotly.NET and Dash.NET):

> So, where does that leave XPlot? Unsure, since the plotly package is the most-used and the most feature complete, and I'm likely to encourage people to move to Plotly.NET once it reaches 2.0.0.

That sounds like a good plan. Contributions from our side follow the same rules as i pointed out above, so help in reaching that 2.0.0 milestone is definitely wanted. I think the strength of XPlot also comes from it offering additional charting APIs. Becoming feature complete in Plotly.NET is ultimately a question of metaprogramming, as it is impossible for me alone (or a small group of OSS contributors) to keep up with the changes coming from plotly.js by implementing them by hand. So maybe tagging @Shmew here, as autogenerating the Plotly Feliz bindings seems to be automated already. Plotly offers a JSON schema for all charts and style parameters, so it is in theory possible to auto generate Plotly.NET bindings from that, maybe via something like Myriad.

@nhirschey regarding static image export: Plotly now uses [Kaleido](https://github.com/plotly/Kaleido) for static image export from charts, which i already wrapped in a quick-and-dirty POC [here](https://gist.github.com/kMutagene/77aca4941e89f96bdbf5700f45ff27b3) but just had not the time to finish and incorporate into Plotly.NET and Kaleido. 

Then there is [Dash.NET](https://github.com/plotly/Dash.NET), which i am currently working on. The python version of Dash is offering jupyter integration, which is also the ultimate goal for the .NET version regarding dotnet interactive. I think when that project matures to this state, we have a pretty nice toolbox for visualization in notebooks.

**Comment by Shmew** [comment]
Yeah [Feliz.Plotly](https://github.com/Shmew/Feliz.Plotly) generates about 95% of the bindings from the JSON schema. Currently it's all done via building strings so I could get something released. If the list style API isn't considered too foreign for normal F# code I think it should be possible to build a library that's implementation independent. We've discussed it a little bit in an issue in the [XPlot](https://github.com/fslaborg/XPlot/issues/90) repo.

**Comment by siavash-babaei** [comment]
Ok, so this has taken a sharp turn towards serious Plotly discussion. Maybe we steer it back towards FsLab and its present and future.

**Comment by dsyme** [comment]
> SciSharp doesn't really have F# on their radar, at least it seems to me that way.

I have discussed this with Haiping, who is one of he technical originators and buiding forces behind SciSharp.

He is very open to aligning F# work into SciSharp, by 

1. adding F# samples/documentation
2. changing the SciSharp messaging to be about “C# + F#” or "F# + C#" as documentation becomes available.
3. bringing some F# community projects under that umbrella where appropriate 

He has made me co-admin on https://github.com/SciSharp to help facilitate changes. Of course I can't make all these changes myself - but it helps in adjusting messaging, getting multiple perspectives etc.

So basically the door is open on the SciSharp side to work towards merging some efforts here towards a win-win for .NET, F# and C# in this space.  

This of course includes continued cooperation and communication with other projects both in and out of the SciSharp umbrella.  For example, I'm working on DiffSharp, which is not under the SciSharp umbrella.  

This is not a full solution, but it gives us options which change the status quo.

**Comment by matthewcrews** [comment]
I'm all for merging efforts with SciSharp. My main concern will be whether the SciSharp libraries will have APIs that are F# friendly. One of the pain points of ML.NET has been that the API does not play well with F# paradigms, at least that has been my experience. I believe wrapping OO method calls in F# functions is frowned upon so I am unsure of the best way to address the style differences of C# and F#.

**Comment by panesofglass** [comment]
> * Julia offers python interop

Both Julia and Swift offer what appears to be very intuitive Python interop. I think something similar in F# would be extremely compelling for those looking for a better alternative. [TensorFlow is betting its future on Swift](https://github.com/tensorflow/swift/blob/master/docs/WhySwiftForTensorFlow.md). That seems to indicate the momentum away from Python is building.

I also echo @lqdev in thinking there should be a focus on Apache Arrow. Could Deedle be built on top? Making it easy to stream Arrow to the browser, especially given ASP.NET's focus and apparent dominance in web performance might push people to .NET for deploying web-hosted models, especially where performance is critical. To that end, an implementation of [Apache Arrow Flight](https://arrow.apache.org/blog/2019/10/13/introducing-arrow-flight/) in F# could be a compelling reason for people to adopt F#.

**Comment by panesofglass** [comment]
> Of course I can't make all these changes myself - but it helps in adjusting messaging, getting multiple perspectives etc.

Even specifying updates, extensions, or changes would be a good first step.

I also noticed that several of the SciSharp projects don't appear to have much progress. I'm thinking specifically of https://github.com/SciSharp/scikit-learn.net.

**Comment by panesofglass** [comment]
What about `query` expressions over data frames, Arrow, or ? This continues to be a recurring theme in many different channels I've tried to follow. There's [Spark SQL](https://spark.apache.org/sql/), [Presto](https://prestodb.io/), [Azure Data Lake Query Acceleration](https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-query-acceleration), a suggestion for an [Arrow Query Engine](https://docs.google.com/document/d/10RoUZmiMQRi_J1FcPeVAUAMJ6d_ZuiEbaM2Y33sNPu4/edit), and many, many more. Is this something that could be handled efficiently through one or more query expression providers?

**Comment by siavash-babaei** [comment]
Dear @panesofglass, F# had a great interop with R through RProvider and Deedle.RPlugin with the excellent feature that data frame from both Deedle and R were compatible and you could almost seamlessly pass around data structure between R and F# and code in R within F# easily. The problem with that typeprovider i suppose is being stuck on .NET 4 and being incompatible with R versions past 3.4 series. I think updating, fixing, and improving an already great existing product, is more logical than starting from scratch. I think there was a plan for a Python typeprovider similar to RProvider but it never really took off. However and whatever the choice, be it R or Python, we require something to cover the significant gaps in F# data tool chain  that will be there for a foreseeable future. Frankly, I doubt anything even Python could achieve R's dominance in terms of number of libraries that can handle any task under the sun, latest algorithms and updates, etc. So it is essential to have interop with R and/or Python.

**Comment by dsyme** [comment]
Just to mention I am looking for collaborators to help work in the SciSharp repositories to make SciSharp a happy place for F# developers.

This is a long journey, but some specific items:

* trialling the use of each major repository with F#

* adding F# samples/documentation

* changing the SciSharp messaging and getting started material to be about “C# + F#” or "F# + C#" as documentation becomes available.

* bringing some F# community projects under that umbrella where appropriate

* adjusting the FsLab documentation to reflect the changes

As co-admin on SciSharp I can make sure this work is efficient and expedited.

Let me know what you think

**Comment by siavash-babaei** [comment]
Are we seeing a pattern forming around various comments as in what components should be in FsLab and in what form:
0- An FsLab Stack akin to SAFE Stack to handle data analytics workflows, is popular. 
1- Need for improved facilities for data ingestion like FSharp.Data on steroids. 
2- Everyone seems to enjoy Deedle. Personally, I love it - thank you @zyzhu, @tpetricek, .... However it has serious limitations in that things that are supposed to know about it as in stuff that is supposed to feed data into it and gizmos that are supposed to get data from it, don't know about Deedle. Frankly, with a bit of work and some additional tools like decent RegexProvider, string manipulation facilities, an statistical package, better Math.NET Numerics, etc. it can admirably emulate the R Tidyverse to some extent. Update and improve Deedle, mayble adding Arrows, and make it compatible with whatever comes on either side, both the data provider feeding into Deedle DataFrame and whatnot, and consumers like FSharp.Statistics or ML.NET, or SciSharp. NOTE: SciScarp has a port of Pandas and that makes Deedle sort of redundant.
3- Plotly.NET appears mature and for visualization seems to be the consensus. I noticed an interesting project DASH.NET which if feature complete, would fill in similar role as R Shiny or Python Bokeh, essential enabling you to bypass commercial software like Power BI or Tableau, allowing a complete data analytics workflow to be undertaken in F#, along with markdown and notebooks and couple of great IDEs, we would have a complete toolset for communication and presentation. NOTE: Matplotlib on SciSharp makes Plotly redundant.
4- We got Math.NET Numerics for linear algebra and some stats. But we need a good one that will provide stuff like tables of descriptive statistics, ANOVA tables, and alike. FSharp.Statistics appears a strong candidate. NOTE: SciSharp had NumPy, Pandas, and SciKit‐Learn making Math.NET and FSharp.Statistics redundant.
5- A complete SciSharp Stack and ML.NET in addition to SPARK.NET would cover like 95% of whatever modelling tasks - in-memory/cloud/etc. Apparently, people find the syntax odd and difficult to use from an F# perspective - too OO/C#-ish. So we need to figure out a way to provide F# syntactic sugar on top of whatever is picked for this bit of FsLab.
6- R and/or Python interop is an essential feature. Maybe revive RProvider, incidentally part of current FsLab. 

There are a couple of great libraries in R Tidyverse: caret for model validation and broom for model representation. They make prototyping much much easier and faster. Similar features in F# would be of paramount importance.

**Comment by siavash-babaei** [comment]
@panesofglass : Query Expressions over Arrows, Data Frames, Data Cubes, SQL and NoSQL databases, etc. Wouldn't that be a killer feature that would bring smiles to many ...

**Comment by siavash-babaei** [comment]
@nhirschey : F# is good for research if you are developing algorithms from scratch. Even Python is still quite some way off R in terms of sheer number of problems it can readily address using a existing libraries, and it always get the latest models and algos and updates, first. That why I think having easy access to R and Python is essential in F#. We cannot hope to emulate such an extended eco system that has been growing for almost 30 years now. 

For me, if I can do the following quickly, seamlessly, and with relative ease, I would judge it to be a decent tool for data analytics. 
1- I am able to read in data from all mainstream sources.
2- I have well-known data structures like data frames by which I can pass around data and input/output across my pipeline.  
3- I can tidy my data and apply transformations, perform selection, join, etc.
4- I can easily process raw string data and easily utilize regular expressions.
5- I can do linear algebra, compute descriptive statistics of all sorts,   generate random numbers, pdf and cdf values, estimate distributions - be it univariate or multivariate or copulas or extreme-value distributions, etc.
6- I can quickly get figures and plots to study relations between variables possibly with plots in matrix format. 
7- I can do the most widely used ML algorithms for linear/ non-linear regression, classification, clustering, and decision trees. Things like multiple linear regression, logistic regression, probit, LDA, baggi.  Mbg and boosting, random forests, SVM, SARIMA, and GARCH, ... probably a couple dozen would solve like 90% of needs.
8- I can utilize concurrency and parallelism to expedite computations. For expensive tasks, I can easily utilize GPU cores as well as CPU cores and I can minimally adjust my code to use a specific optimization methods or cloud resources for computations.
9- I have access to all mainstream platforms for handling big data like Spark, and for high-end modelling as in deep learning and similar, like Tensorflow and Keras, etc.
10- I have access to excellent static and interactive visualization facilities, along with markdown and notebooks to collaborate, communicate, and present. A tool to handle building dashboards and reports and publish makes the experience full.
...
13- Based on size of my data and resources and whatnot, I have the toolset that I need to do the task in-memory or over a distributed local cluster or in-cloud - with hopefully consistent syntax across libraries requiring minimal code change, or at least preserving the solution logic to greatest extent.

**Comment by muehlhaus** [comment]
This is a very interesting thread, and I am happy about the lively discussion.
I am teaching F# as a first language for data science in bioinformatics since couple of years now, and I am very fascinated how fast students are able to do data science with F#. It is just a fantastic language for this purpose.
I like the idea to join forces under a .NET data science umbrella (SciSharp), if I understood @dsyme correctly. 
However, I think it is necessary to provide a haven for F# first developers in data science. E.g. for students, who just learned F# as their fist language and are very proud of their achievements, it might be hard to comprehend that they should implement in C# and build thin wrapper around it…
Additionally, I think that libraries written in pure F# can also benefit from the language paradigm. There are excellent examples like diffSharp or flips  (@matthewcrews ) and many more…
This is the reason why we started to develop libraries in F# (BioFSharp, FSharp.Stats, Plotly.NET) and believe this could be a complementary approach.  
Me and my group would be willing to put some efforts into it. Maybe we can discuss it further.

**Comment by siavash-babaei** [comment]
@muehlhaus I think the idea behind adopting something like SciSharp is: 1- Get coherency and consistency in code, 2- Add F# wrapper on top of existing stack which is certainly much faster and doable than natively developing all that from scratch. 3- Gain enough adoption and foothold 4- use user-base to affect development of underlying stack, get native bindings, etc.

I mean we have tried doing it this way for 15 year and despite all niceties of F#, we are where we are. Maybe we accept current limitations, optimize efforts and approach it differently. Specially when you have assets like @dsyme that as he just did, can bring F# to the attention and care of SciSharp team.

**Comment by muehlhaus** [comment]
@siavash-babaei  There are two different paths that do not seem to me mutual exclusive. To be productive quickly building F# wrappers is a very good idea. However, the fact that there is a linear regression implemented in C did not stopped a python programmer to implement one in pure python.  
For us being able to extent a library in our favorite language is of some importance. I don’t know if you understand my concerns.

**Comment by muehlhaus** [comment]
@dsyme  Do you think SciSharp would be a good place for F#-first projects like e.g. FSharp.Stats or is there a different philosophy needed?

**Comment by siavash-babaei** [comment]
@muehlhaus not only i do not think them mutually exclusive but mutually beneficial.  Your example of Python is actually good here: for quite a while pythonistas relied on wrappers around existing code in say C, moreover even now many libraries of R and Python rely on bits and pieces of code in Fortran/C/C++. You are also ignoring important non-technical aspects like size/momentum and variety-of-expertise of user base, corporate support, business logic, etc. The size, composition, and expertise of F# community - especially when it comes to statistics/data analytics/data science - is such that it makes developing a large cohesive project slow and difficult often without decent documentation and training material. Furthermore, maintenance becomes so dependent on few individuals that quite often we get great projects like mbrace or rprovider or whatever that kind of go into permanent comatose status. Python has millions of active users from research scientists at MIT and Microsoft to data analysts at small firms covering all aspects of subject and vast corporate support, so it can afford to do whatever the heck it wants. As example, the slack channel for F# Data Science has like 500 members, while the Datacamp R Learning channel has something like 20,000 members. 
From a business and productivity perspective under current circumstances and for data science, it makes every sense to focus most efforts and very limited resources into providing a toolbox that can be used to address majority of problems, take it from there with growth and maturity. Once passed a certain threshold, then it would be not only viable but mandatory to not only develop native tools, but move past that and innovate turning projects like mbrace into killer features - like Spark did for Scala - things that Python and R would adopt and add bindings for.
Say there are 1,000 people that can spend an average of 200 hours per year contributing to F# data science. Starting now, what can you achieve when you spend 200,000 hours on 20 problems some of which are partly or wholly solved in a compatible language, versus trying to solutions from scratch for 100 problems. 
As a bonus point, suppose the platform you are adopting from C# has good documentation, samples, and training material; just imagine how much easier it would be to also adopt documentation and tutorials. This is by the way, an important pain point for F#: you want to do something in C#, in many cases you are bound to find really good documentation and text/video tutorials that get you where you want in short order.

**Comment by dsyme** [comment]
> @dsyme Do you think SciSharp would be a good place for F#-first projects like e.g. FSharp.Stats or is there a different philosophy needed?

That's a good question. Tricky. Worth a good think before acting.

SciSharp is effectively a machine learning stack  - the strongest parts are TensorFlow.NET, Keras.Net, Numpy.NET (I may have missed some). That's a good and important focal point and I think we don't want to try an F# replacement for those (DiffSharp can be seen as additive/overlapping and is very much in line with these, and could one day move into SciSharp).

So right now I'd say "no" or at least "not yet" for F#-specific projects there.  We're slowly tweaking C# projects to be joint F#/C# and it would not yet be right to bring F# specific things there I think

So we haven't yet really solved the question of where an F#-first home for data science assets lives and a place for the community to congregate.  Giraffe (web), Ionide (tools), Fable (web), Fabulous (native mobile) and "fsprojects" (big bucket) all offer some inspiration.  For some strange reason it just didn't work for "fslaborg".

FSharp.Stats is absolutely great.   @muehlhaus If you're looking for a new home for it, I'm wondering if you would actually like to consider taking a lead in rejuvenating https://github.com/fslaborg (or another org?).  By this I mean you'd become and admin here:

1. completely stripping away the existing website and replacing it

2. moving FSharp.Stats in here.

3. actively encouraging new F#-specific data-science projects to move in.

4. still embracing SciSharp as an aligned partner org for C#/F# projects and machine learning in particular.

FSharp.Stats does feel like an opportunity to start to re-create a centre of gravity

**Comment by siavash-babaei** [comment]
@dsyme: What about Deedle, FSharp.Data and other working/up-to-date TypeProviders, Plotly.NET, etc. along with FSharp.Stats, as centre of gravity!?

**Comment by dsyme** [comment]
> @dsyme: What about Deedle, FSharp.Data and other working/up-to-date TypeProviders, Plotly.NET, etc. along with FSharp.Stats, as centre of gravity

A good "centre of gravity" is a place where likeminded, empowered people come to collaborate, create and deliver over a long period of time with some cooperation and some creation.  "fsprojects" and "Giraffe" and "Ionide" have had this role.  "FsLab" has not, at least not sufficiently so far.

The experience with FsLab has shown how easy it is to create a "centre of gravity" that goes stale and where major development sort of grinds to a halt as tech changes (e.g. the change from .NET Framework to .NET Core).  Probably the very worst part of this was the attempt to create an "integrated uber-package" called "FsLab" that was the one-stop integration of a set of packages.  It sounds so nice in theory but in practice it went stale very fast - it only took one of those packages to stop development or have a Windows dependency to basically grind the whole thing to a stop. Also, centres of gravity can interact badly with each other, and above all they need regular spring-cleaning and rejuvenation.

Probably the way forward is to rebrand FsLab along the lines of fsprojects - "an F# community incubation space for data science and related projects" and just let many flowers bloom.

**Comment by siavash-babaei** [comment]
@dsyme thank you for clarifying what FsLab is going to be: An incubation environment for F# Data Science projects, where projects will come and go based on usability, adoption, maintenance,etc. COOL

**Comment by muehlhaus** [comment]
@dsyme This sounds very good and exciting. I discussed it with my group, and we would be very happy to follow this path thereby taking responsibility. We believe an incubation space for F# first data science projects is very important and FsLab would be an ideal home for FSharp.Stats.

Therefore, we would like to move FSharp.Stats there and become active member of FsLab. 

Does that sound good to you?

**Comment by siavash-babaei** [comment]
@muehlhaus thank you - and your team - for taking on the leading role. When you get around to it, I am sure quite a few people would be interested in your ideas and plans, etc. even the just of your brainstorming sessions ...

**Comment by siavash-babaei** [comment]
Also, simple guide-lines and how-to documents for contributions at all levels both for FsLab and SciSharp. I think it would help people with all levels of expertise in various fields to make whatever small or large contributions they can make. With a smaller community, squeezing every resource as much as possible, would be even more vital.

**Comment by dsyme** [comment]
@muehlhaus That is fantastic news, and I'm very glad it has the backing of your team as well.

I'll make you admin for the org - feel free to bring on others.  I'll also strip away the existing website to give minimal positioning as a community incubation space to allow us to start with a cleaner slate

**Comment by dsyme** [comment]
Some progress:

1. I have taken down the old website, the Razor/paket/... source for it is here https://github.com/fslaborg/zzarchive-fslaborg.github.io

2. The new website at it just currently a generic display of the projects in the incubation space.  @muehlhaus feel free to populate with alternative material as things come together using whatever web tech you like :-)

3. I will rename this project and archive it at https://github.com/fslaborg/zzarchive-FsLab.  (This project served as the "master" project for the FsLab nuget uber-collection and package with data scripting (much of which has been moved to fsdocs), but I think we should retire that whole collection of stuff to make way for the new.  The existing package will be available)

4. The current projects in the incubation space are

       https://github.com/fslaborg/XPlot
       https://github.com/fslaborg/Deedle
       https://github.com/fslaborg/FSharp.Charting
       https://github.com/fslaborg/RProvider

FSharp.Charting is windows-only and should be renamed and moved to fsprojects - we don't want *any* windows-only stuff in FsLab going forward.

XPlot will probably soon be better replaced by Plotly.NET

So this is really Quite An Empty Space.  Let's fill it with goodness and make sure to also embrace other assets like FSharp.Data, Giraffe etc. which live outside this space.

**Comment by dsyme** [comment]
This discussion is continues at https://github.com/fslaborg/FsLab/discussions/3

Please join us there!
